Windows: DeviceTree traversal/offset bugfixes

dgmcdona opened the pull request:
There are two issues with the current implementation of the `windows.devicetree.DeviceTree` plugin: 
- The traversal of the device tree is incomplete. The `DEVICE_OBJECT.get_attached_devices` extension method does not traverse the full breadth at each level of the device tree through the `NextDevice` linked-list.
- Every entry uses the driver's virtual address as the offset. Device entries should yield their own virtual addresses in the generator.

This PR fixes the issue by implementing a recursive, private staticmethod on the `DeviceTree` plugin class that does a complete depth-first traversal of the tree, yielding many more device objects (using their virtual address as the `Offset` member instead of that of the parent driver).

dgmcdona committed:
Windows: Use device offsets instead of driver offsets

This updates the `devicetree` plugin to return the virtual offsets of
the device objects themselves instead of that of their parent driver.

dgmcdona committed:
Windows: Fix devicetree traversal

Previously, the attached devices were being traversed once depth-wise,
but the horizontal levels of the device tree were not being properly
traversed, leading to tons of missing device entries. This fixes the
issue by implementing a recursive staticmethod for properly descending +
enumerating devices on each level of the tree.

ikelos commented on volatility3/framework/plugins/windows/devicetree.py:
```diff
@@ -186,6 +147,48 @@ def _generator(self) -> Iterator[Tuple]:
                 )
                 continue
 
+    @staticmethod
```
Please don't use staticmethod inside versionable objects (even for private methods) because it prevents us being able to get back to the version from just the method.

ikelos reviewed (commented):
Looks ok, but please no static methods in classes that are versionable.  Also, I'd tend to agree with the scanner and suggest that at least a comment explaining why the exception can be ignored should be included...

dgmcdona commented:
@ikelos One consideration for this should be what to do about the `get_attached_devices` method on the `DEVICE_OBJECT` extension class. It's current implementation is flawed, which is why I moved away from it in this PR, but maybe we should deprecate + remove or fix it as part of this effort. I would have fixed it and used it here but it would have required a method signature change in order to also return the depth in the device tree that we need when constructing the `TreeGrid`.

dgmcdona committed:
Windows: Add log statements on InvalidAddressExceptions

dgmcdona committed:
Windows: Cleanup devicetree + extension class

Moves a lot of the code that was parsing information about device
objects out of the `DeviceTree` plugin logic and into the
`DEVICE_OBJECT` extension class. Also moved the dict mapping integer
values to device types into the extension module.

Fixes a bug within the extension class' `get_attached_devices` method -
now, immediate children of the device in the tree are yielded, instead
of the faulty deep traversal that was occuring before. Because of this
bugfix in an extension class as well as the addition of new methods in
the extension class, this also bumps the framework minor version number.

ikelos commented:
I'd pick a different method name, and run them alongside each other (deprecating the old one)?  Eventually, if it's really worth it, you can shift the new one back to the old name.

If the code belongs centrally, it's much better to have it live there than in a plugin where someone may not know about it and try to reimplement (possibly incorrectly).  If it's only for the generator (ie, not to be used by other plugins) then it's fine.  You could also consider a visitor design pattern, that takes a function and runs it on each node in the tree in a particular order, calling it with the node and the depth at each point?

dgmcdona commented:
Thanks, marking as draft again while I revisit this.

ikelos commented:
I didn't think this was too far off prime time?  Could you dust it off and see what would be needed to get it flying again please?
